controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -634,75 +734,183 @@ private long findSegmentSplitsMerges(List<StreamSegmentRecord> referenceSegments
+                                    }
+                                } else {
+                                    // if segment is predecessor of another segment in the stream cut then it has to be 
+                                    // fully read 
+                                    fullyReadFrom.add(segmentRecord);
+                                    // find successors of segmentRecord and add it to tocheck list
+                                    int segmentEpoch = NameUtils.getEpoch(segmentRecord.segmentId());
+                                    int index = segmentEpoch - epochLow;
+                                    for (int i = index; i < epochs.size(); i++) {
+                                        if (!epochs.get(i).containsSegment(segmentRecord.segmentId())) {
+                                            epochs.get(i).getSegments().forEach(x -> {
+                                                if (x.overlaps(segmentRecord)) {
+                                                    toCheck.add(x);
+                                                }
+                                            });
+                                            break;
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+
+                    toCheck.remove(segmentRecord);
+                }
+            }
+        }
+        return isValid;
+    }
+    
+    private List<StreamSegmentRecord> findSegmentsForMissingRange(EpochRecord epochRecord, Map.Entry<Double, Double> missingRange) {
+        return epochRecord.getSegments().stream().filter(x -> x.overlaps(missingRange.getKey(), missingRange.getValue()))
+                          .collect(Collectors.toList());
     }
 
     private CompletableFuture<Boolean> isStreamCutValidForTruncation(Map<Long, Long> previousStreamCut, final Map<Long, Long> streamCut) {
         if (previousStreamCut.isEmpty()) {
             return isStreamCutValid(streamCut);
         } else {
             return isStreamCutValid(streamCut)
-                    .thenCompose( isValidStreamCut -> {
+                    .thenCompose(isValidStreamCut -> {
                         if (isValidStreamCut) {
-                            return computeStreamCutSpan(streamCut)
-                                    .thenCompose(span -> computeStreamCutSpan(previousStreamCut)
-                                            .thenApply(previousSpan -> greaterThan(streamCut, span, previousStreamCut, previousSpan)));
+                            CompletableFuture<ImmutableMap<StreamSegmentRecord, Integer>> span1 = computeStreamCutSpan(streamCut);
+                            CompletableFuture<ImmutableMap<StreamSegmentRecord, Integer>> span2 = previousStreamCut.isEmpty() ?
+                                    CompletableFuture.completedFuture(ImmutableMap.of()) : computeStreamCutSpan(previousStreamCut);
```

Comment:
`span2` will never be equal to `CompletableFuture.completedFuture(ImmutableMap.of())` becuase if `previousStreamCut.isEmpty=true`, that case is already handled in the if block.